### ISSUE TITLE
Huge memory use

### ISSUE THREAD

**Issue by perry-mitchell** [meeb/tubesync#428]
Hello,

I've been debugging my Unraid instance lately, as it's been crashing since I added the TubeSync container on it. I've added some small channels that work fine (~150-200 videos), but adding a 5K+ channel like Markipilier or LTT causes huge amounts of memory to be chewed up.

![Screenshot from 2023-10-24 17-11-31](https://github.com/meeb/tubesync/assets/3869469/cf06b0b0-fa91-4ccc-ad6f-b41ad4a30445)

I found out my system was probably running out of memory, due to it being unconstrained:

![Screenshot from 2023-10-24 16-51-03](https://github.com/meeb/tubesync/assets/3869469/845c4728-a5e8-495e-8524-1b1bd116575d)

So I followed the directions in [this issue](https://github.com/meeb/tubesync/issues/103), and locked the memory to 4G. But now when I try to index any channel, big or small, I get:

```
[2023-10-24 18:20:38 +0300] [319] [ERROR] Worker (pid:365) was sent SIGKILL! Perhaps out of memory?
```

Mind you, all channels have been scraped already.. so it shouldn't be doing as big a job again. I'm also using an external MariaDB instance.

Any ideas what I can test next?

EDIT: Extra context:

![image](https://github.com/meeb/tubesync/assets/3869469/e195cae4-b63c-4d4a-8ec3-0531d699e2d4)

![image](https://github.com/meeb/tubesync/assets/3869469/e83d72ba-30cc-4389-9076-541935a5be25)

```
2023-10-24 17:17:18,662 [tubesync/INFO] Successfully downloaded media: kkEypAe3rpE (UUID: c0fb92b0-fb6e-49da-903b-673074e79c34) to: "/downloads/video/shiey/2023-09-30_shiey_mission-to-enter-and-escape-collapsing-mine-tunnels_kkEypAe3rpE_1080p-vp09-opus-60fps.mkv"
2023-10-24 17:17:18,702 [tubesync/INFO] Copying media thumbnail from: /config/media/thumbs/c0/c0fb92b0-fb6e-49da-903b-673074e79c34.jpg to: /downloads/video/shiey/2023-09-30_shiey_mission-to-enter-and-escape-collapsing-mine-tunnels_kkEypAe3rpE_1080p-vp09-opus-60fps.jpg
2023-10-24 17:17:18,703 [tubesync/INFO] Writing media NFO file to: to: /downloads/video/shiey/2023-09-30_shiey_mission-to-enter-and-escape-collapsing-mine-tunnels_kkEypAe3rpE_1080p-vp09-opus-60fps.nfo
2023/10/24 18:20:38 [error] 347#347: *156 upstream prematurely closed connection while reading response header from upstream, client: 192.168.0.50, server: _, request: "POST /tasks-reset HTTP/1.1", upstream: "http://127.0.0.1:8080/tasks-reset", host: "192.168.0.100:4848", referrer: "http://192.168.0.100:4848/tasks-reset"
[2023-10-24 18:20:38 +0300] [319] [ERROR] Worker (pid:365) was sent SIGKILL! Perhaps out of memory?
[2023-10-24 18:20:38 +0300] [838] [INFO] Booting worker with pid: 838
2023-10-24 18:20:39,336 [tubesync/INFO] Using database connection: django.db.backends.mysql://tubesync:[hidden]@MariaDB:3306/tubesync
2023-10-24 18:20:43,353 [tubesync/INFO] Using database connection: django.db.backends.mysql://tubesync:[hidden]@MariaDB:3306/tubesync
```

EDIT 2: I can see the kernel killing the processes at the unraid/OS level:

```
Oct 24 18:20:41 Nostromo kernel: oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0,oom_memcg=/docker/c6047e9c67bcbc030ea0cd7f4ef029403aed74d4b481f58b3cdb480a04b47464,task_memcg=/docker/c6047e9c67bcbc030ea0cd7f4ef029403aed74d4b481f58b3cdb480a04b47464,task=python3,pid=15412,uid=99
Oct 24 18:20:41 Nostromo kernel: Memory cgroup out of memory: Killed process 15412 (python3) total-vm:4294580kB, anon-rss:3814516kB, file-rss:0kB, shmem-rss:0kB, UID:99 pgtables:7956kB oom_score_adj:0
```

**Comment by perry-mitchell** [comment]
Seemed stuck again, so I reset the tasks again.. Memory quickly climbed to my new limit of 6G and it died again:

```
[2023-10-24 18:43:19 +0300] [319] [CRITICAL] WORKER TIMEOUT (pid:364)
2023/10/24 18:43:19 [error] 346#346: *24 upstream prematurely closed connection while reading response header from upstream, client: 192.168.0.50, server: _, request: "POST /tasks-reset HTTP/1.1", upstream: "http://127.0.0.1:8080/tasks-reset", host: "192.168.0.100:4848", referrer: "http://192.168.0.100:4848/tasks-reset"
[2023-10-24 18:43:19 +0300] [364] [INFO] Worker exiting (pid: 364)
192.168.0.50 - - [24/Oct/2023:18:43:19 +0300] "POST /tasks-reset HTTP/1.1" 502 150 "http://192.168.0.100:4848/tasks-reset" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/118.0"
[2023-10-24 18:43:20 +0300] [319] [ERROR] Worker (pid:364) exited with code 1
[2023-10-24 18:43:20 +0300] [319] [ERROR] Worker (pid:364) exited with code 1.
[2023-10-24 18:43:20 +0300] [432] [INFO] Booting worker with pid: 432
2023-10-24 18:43:20,206 [tubesync/INFO] Using database connection: django.db.backends.mysql://tubesync:[hidden]@MariaDB:3306/tubesync
```

**Comment by d0ugal** [comment]
How many workers do you have?

**Comment by perry-mitchell** [comment]
>How many workers do you have?

Sorry I didn't see this. I've since removed my copy of Tubesync while I plan on migrating it to a new system. So I don't recall exactly what I had configured there. I suspect it was 2 workers, but could have been 4 :man_shrugging:.

Imo there shouldn't be any need to use so much memory - Playlists and channels can be huge so if pulling this data is what's causing the huge use of memory, perhaps it should be written to disk instead of expecting the host to have enough memory to handle any channel size. Databses like MySQL do this - not sure what storage mechanism Tubesync uses under the hood but if it were a DB like this it should rollover onto disk just fine.

**Comment by meeb** [comment]
There isn't any need for that much memory and it's absolutely a bug. The data collected by the worker can be huge in some cases and that's stored in RAM, however it's not being effectively garbage collected which is related to how the worker library is or isn't working properly. As for temporary spooling of large data to disk, that would be a yt-dlp feature and not a tubesync feature.

**Comment by perry-mitchell** [comment]
Thanks for clarifying! To be clear, is the garbage collection bug on tubesync's side or yt-dlp's?

I'll be setting up a new node probably in the next week or so.. If there's more debugging information I can provide, in case this happens again, please let me know and I'll do my best to share it here.

I'll be running it in an LXC next though, instead of just Docker on Unraid, so the environment will be subtly different.

**Comment by perry-mitchell** [comment]
> The worker library in tubesync has been on the cards to be replaced for... some time now, it's just a lot of work and I've not had time.

Understandable. Anyway, I really appreciate the project so I'll keep on eye on this going forward :)

**Comment by locke4** [comment]
I can reproduce this issue also. I'm working on a branch this weekend which I'm hoping I can come up with a short-term fix. Only saw the issue when I was running some VMs, which reduced the memory available on my server. I'm hoping setting max_requests_jitter and max_requests on the gunicorn workers might improve this and force each worker to give up it's memory allocations when a large job is being processed by them.

**Comment by meeb** [comment]
@locke4 this isn't going to be anything to do with gunicorn, it's the background workers which are far more complex to replace.

**Comment by bawitdaba** [comment]
I've also been having this issue, though I do use TubeSync a little differently then most. I have a single Plex Watch Later playlist that I manually sync in TubeSync when I add new items, and I also use TubeSync with MariaDB and I have 64 GB of RAM. Eventually it gets to the YouTube playlist limit of 4,000 videos and becomes unmanageable in TubeSync.

When I max out or near max out the YouTube playlist TubeSync will use an ungodly amount of system RAM easily over 16 GB of RAM and eventually all of it until the system becomes unresponsive.

I tried creating a new empty playlist and pointing TubeSync at that, and while that did allow the indexing to go quicker Docker would still kill the python3 process for using over 8 GB of RAM the new limit I imposed on the container.

Even though the playlist is empty, all the old playlist items are still stored in sync_media. Which this was expected, but what I noticed was that django is still loading all of these items and taking a ton of ram to do so. I also saw this when loading the /admin panel as well.

I emptied the tables sync_media and background_task_completedtask and TubeSync becomes usable again taking only a few hundred MB of ram.

**Comment by locke4** [comment]
> @locke4 this isn't going to be anything to do with gunicorn, it's the background workers which are far more complex to replace.

Fair enough, you're talking about the django background_task as the background workers? I'm running a test on a channel which I know breaks tubesync:latest but assuming it won't have any improvements if it's not relating to the gunicorn or celery workers. I have a fix relating to the filter_text behaviour for this PR anyway so good to try and get that completed this weekend regardless.

I want to see if I can look into a fix short of replacing them as I appreciate replacing is a much larger task. For extra info, RAM allocated during indexing also doesn't seem to be dropped which exacerbates this (below is an indexing task which completed around 8am but memory never released, deleting the channel immediately crashes the docker container with SIGKILL out of memory errors). 11am is me updating to my branch and repeating the test.

![image](https://github.com/meeb/tubesync/assets/65832338/849c3449-8d7f-4f3b-8fca-693ca4559c18)

**Comment by locke4** [comment]
As expected, doesn't fix the excessive memory usage during a delete operation. I was however able to complete the delete task (though the web browser timed out, the container didn't crash out with a SIGKILL error) which I don't really understand. I found https://github.com/arteria/django-background-tasks/issues/186 which might be promising but forking django-background-tasks isn't the best solution long-term.

![image](https://github.com/meeb/tubesync/assets/65832338/bbd68db8-3a5c-45c1-8855-8b15d408c60e)

**Comment by perry-mitchell** [comment]
Migrated my setup from Unraid to Proxmox, now starting off with a smaller 300 video channel. Seems ok so far:

![image](https://github.com/meeb/tubesync/assets/3869469/10dd5150-af8a-4c79-b23e-f8200cb5ee9b)

Once this is done I'll try again with another 5K video channel, see how it likes that.

**Comment by perry-mitchell** [comment]
Also, even once a source successfully syncs, I can't open it without it crashing: 
![image](https://github.com/meeb/tubesync/assets/3869469/fac86e45-3965-452a-ace1-91db9834daf8)

Memory climbs to 2-3GB more, then this crash.

![image](https://github.com/meeb/tubesync/assets/3869469/7ad386b7-6123-423d-b2fb-4342a5d13e64)

Before:

![image](https://github.com/meeb/tubesync/assets/3869469/3aeb4c0c-3926-47aa-b7b6-09937cb26103)

During:

![image](https://github.com/meeb/tubesync/assets/3869469/57acfb50-5567-4afa-80e5-7544855a7952)

**Comment by meeb** [comment]
@locke4 Yes, that's the plan. However I should warn you that this is a _significant_ amount of work. General rewrite and refactoring big. Porting the tasks to celery tasks is the easy bit. Migrating all existing tasks (and task management and UI updates on tasks) without breaking anyone's installs is going to be much harder.

**Comment by perry-mitchell** [comment]
Putting my hand up now to be a nightly build tester ✋ 😁

**Comment by meeb** [comment]
Just to mention here that even with rewriting the worker and tasks system this won't really reduce the peak memory requirement much, it'll just mean it gets freed again after use.

**Comment by locke4** [comment]
I've got a couple of thoughts that'd I'd appreciate your thoughts on @meeb after my prototyping my feature branch on this issue yesterday. 

1. I've created celery tasks for all web interactions so the user immediately gets a success response, resolving the gunicorn timeout which is related to very large channels. Background-tasks is still used but moved into each celery worker thread. Not only does this massively simplify the task, but it should make this a non-breaking change and all the existing UI still works.
2. I've split the most horrifying loop, the media.save loop. The loop is now performed in the celery worker thread and the media.save creates a celery task each time the loop runs. This had a huge performance increase as previously a single gunicorn thread was running the task, now it's multithreaded by the number of celery worker threads. Looked like between 5-10x speedup in tasks like updating a source or deleting a source in the backend.
3. I haven't seen any SIGKILL errors yet. It looks like the memory requirement is essentially throttled by the number or celery threads.

I've got a couple of weird interactions I need to solve before I commit this feature branch to my fork repo but generally I think this approach is a good candidate fix. I should have a PR ready in a week, provided you're happy with this approach

**Comment by meeb** [comment]
Thanks for the effort, @locke4! Specifically for your questions:

1. Ideally, `background_tasks` needs to be completely removed. This is blocking upgrading Django beyond 3.2 as well as nesting tasks is going to cause unexpected issues with scheduling and likely cause other more complex problems. This was a major reason it's taken so long. A dual celery/background_tasks implementation sounds pretty nightmarish. Offloading all work to the background is excellent though and the general end game!
2. Sounds reasonable, however this likely needs locks to prevent race conditions as well as other less common situations. Note anyone using an SQLite backend (which is most people) can only have a single concurrent write and the `yt-dlp` interacting worker should only run one process at once (any more than that and YouTube starts to get angry) so some concurrency controls are going to be needed.
3. The memory limit is actually just indexing channel metadata from `yt-dlp` which is returned as a massive, massive dictionary for a large channel. There's not much optimization that can be done here, other thank make sure the worker frees it properly once processed (which is what `background_tasks` isn't doing).

Generally I think you're on the right approach, but I would suspect there's going to be a _lot_ of edge cases, weirdness and other issues that will need to be worked out with this transition. I've had two attempts at this myself and never fully finished it enough to merge. If this has any bugs in at all it's going to annoy a lot of people.

**Comment by perry-mitchell** [comment]
@meeb Not to derail, and perhaps this could be another issue elsewhere, but has anyone reached out to the yt-dlp project regarding this metadata? Surely a JSON stream or something incremental is of better design than a massive blob of data?

It's not sustainable for either project, imo, to function like that. It'll still prevent anyone from scanning popular channels unless they're willing to dedicate 3-6GB of RAM for the instance. I'd have hoped to run such a container at 1-2GB max in a perfect world.

Just my thoughts 😊

**Comment by meeb** [comment]
I have not, and quite honestly it might be how I've implemented it rather than an upstream issue. Worth checking both ends though! I hacked up the original TubeSync in a weekend to scratch a personal itch so... there may well be legacy silliness that worked fine for the smaller channels I sync but not for people trying to parse gigabytes of metadata. I've not stumbled over it yet but there may be a streaming metadata feature buried in the yt-dlp API.

**Comment by locke4** [comment]
> Thanks for the effort, @locke4! Specifically for your questions:
> 
> 1. Ideally, `background_tasks` needs to be completely removed. This is blocking upgrading Django beyond 3.2 as well as nesting tasks is going to cause unexpected issues with scheduling and likely cause other more complex problems. This was a major reason it's taken so long. A dual celery/background_tasks implementation sounds pretty nightmarish. Offloading all work to the background is excellent though and the general end game!
> 2. Sounds reasonable, however this likely needs locks to prevent race conditions as well as other less common situations. Note anyone using an SQLite backend (which is most people) can only have a single concurrent write and the `yt-dlp` interacting worker should only run one process at once (any more than that and YouTube starts to get angry) so some concurrency controls are going to be needed.
> 3. The memory limit is actually just indexing channel metadata from `yt-dlp` which is returned as a massive, massive dictionary for a large channel. There's not much optimization that can be done here, other thank make sure the worker frees it properly once processed (which is what `background_tasks` isn't doing).
> 
> Generally I think you're on the right approach, but I would suspect there's going to be a _lot_ of edge cases, weirdness and other issues that will need to be worked out with this transition. I've had two attempts at this myself and never fully finished it enough to merge. If this has any bugs in at all it's going to annoy a lot of people.

Thanks for the comments! On 1., agreed that it's not ideal to have both. At the moment I've been focussed on just fixing the two breaking bugs with large channels which I believe is now complete bar writing test cases and a couple of edge cases.

I'm going to complete this feature branch fully as a fix branch. I'll then work on fully removing background-tasks on a new feature branch, so if I'm unsuccessful at least the fixes can be merged from my current branch and a full refactor completed later building on this. My main issue I haven't solved yet is how to ensure tasks are migrated or recreated in celery on update.

I was thinking that one way of ensuring fully removing background-tasks isn't a breaking change is to force reset tasks on boot in some intelligent way (to avoid an unnecessary task reset).

**Comment by meeb** [comment]
OK, also remember that if you're not tackling the concurrency and race condition issues initially either lock the celery workers to 1 (and disable the ENV var for worker count) as well! Happy to review what you submit as a PR.

**Comment by Leopere** [comment]
Okay I reassessed my settings and in my case I just increased the number of workers from 1 up to 10 because why not.

**Comment by meeb** [comment]
And thanks for the log, that just shows the worker using a load of memory for some reason and getting killed, however every bit helps.

**Comment by Leopere** [comment]
> You should definitely keep your workers at 1. There's no technical reason locally you can't run more workers, however YouTube quickly gets angry at you and blocks your IP or imposes other similar restrictions if you run more than 1 worker.
oh fair enough... good call I'll dial it back to 2 at most then i guess it just seems way more responsive now that I set it higher. 

Thanks for the fast reply @meeb

**Comment by meeb** [comment]
You can ignore the message to run `manage.py migrate` - that shows up just because your download path is always different to the one in the default schema and it thinks you need to update it, you don't.

As for your errors, the first once is that playlist wasn't indexable for some reason, try using `yt-dlp` on the command line with that playlist and see what it returns. Indexable in this case means "tubesync attempted to invoke yt-dlp to get a list of the video IDs in the playlist and this failed". As to why it failed, it can be anything really, the playlist is private and you've not set your `cookies.txt`, the playlist doesn't exist, YouTube has blocked your IP, your internet connection was down, etc. Basically anything that would cause the index request to fail.

The second error is self-explanatory, tubesync doesn't have permissions to create that directory so check the local directory permissions.

**Comment by Leopere** [comment]
I have the playlist as unlisted but it shouldn't be private will that cause a problem I just like telling Tubesync to slurp up video I think could be useful in the future which may get taken down for various reasons.

**Comment by meeb** [comment]
Try a `manage.py youtube-dl-info "https://youtube.com/playlist=... your playlist ID here ..."` and see what is returned. That command will just see if the playlist is indeed index-able.

**Comment by timwhite** [comment]
@locke4 how did you go with moving things to celery? There will be obvious merge conflicts now as you don't appear to have done much in awhile (https://github.com/meeb/tubesync/compare/main...locke4:tubesync:celery-refactor). Was this mostly working, or are there major issues still?

Can we split this up a little and help each other move it forward? Also which is the right branch you've been working on? Thanks

**Comment by meeb** [comment]
@timwhite I'm waiting or django-tasks to be merged in the next Django 5 release then going to drop the idea of using celery entirely.

**Comment by timwhite** [comment]
> @timwhite I'm waiting or django-tasks to be merged in the next Django 5 release then going to drop the idea of using celery entirely.

That's good to know. I assume you mean https://github.com/django/deps/blob/main/accepted/0014-background-workers.rst and something like https://github.com/RealOrangeOne/django-tasks being merged into core?

And I assume the advantages are that it's Django native, instead of adding celery and redis?

Knowing that, I may continue to tweak some existing tasks as is, to try and improve how they run, instead of trying to offload them to Celery. I can see some of the places large channels crash, could be handled better with a task handed off to a worker so the UI can return quicker. Any tips on adding more tasks? I think in particular working on moving some of code from `source_post_save` into a task so saving a source no longer crashes the interface (for large channels).

**Comment by meeb** [comment]
Yeah, `django-tasks` being the reference implementation. Something mainline and not requiring a lot of heavy external libraries would be desirable.

This isn't actually related to memory issues, that's to do with the current workers occasionally not freeing memory after invoking `yt-dlp` (and would be solved by changing the worker implementation, hopefully) which offloading more work to them won't fix. It would make the interface snappier though.

I've just moved a small loop on `source_post_save` to a background task as an example of how to implement the current background task system if you wanted to try to move any more: https://github.com/meeb/tubesync/commit/b6cec8ab1164034a29e36e4173e073f3b696b7f3

The current task system uses https://django-background-tasks.readthedocs.io/en/latest/ however it's now obviously terribly out of date and doesn't support anything above Django 3.2. It was fine years ago when I hacked up the first version well before anyone else ever used tubesync...

Switching to celery or the new `django-tasks` is a much bigger job and will need its own issues to track it. The memory usage issue isn't likely to go away, at its peak invoking `yt-dlp` to index the media on a very large channel is still going to return an absolutely vast JSON object or dict which is always going to use a significant amount of RAM to process. This should spike and then be freed though, not just sat allocated forever in some cases.

You're welcome to attempt to offload other tasks to the current background worker tasks if you want, just be careful of args (which need to be pretty much just strs or ints, serialization isn't particularly advanced) and state race conditions.

All the stuff at the top of `tasks.py` like https://github.com/meeb/tubesync/blob/main/tubesync/sync/tasks.py#L41 is to account for `django-background-tasks` not having a built in way to see what's running so this reverse engineers some of its internals to pull out some running tasks data for the `/tasks` page.

**Comment by FaySmash** [comment]
As a temporary workaround I added 
```
    deploy:
      resources:
        limits:
          memory: 8G
```
as a service attribute to the docker-compose and it runs fine so far. 8GB is still a lot but it completely consumed the 32GB of the VM before \^^
